ando/AnDOChecker.py

```diff
@@ -0,0 +1,209 @@
+                    get_list_of_rules_with(current_mandatoryfolder_rule))
+                for each_file in list_of_mandatory_folders:
+                    dir_res = [re.compile(each_file).search(d) is None for d in dirs]
+                    if all(dir_res):
+                        error_list.append(
+                            "Mandatory folder not found for this rule : {}".format(current_mandatoryfolder_rule))
+        ###
+        # 2. check whether the rules are followed for the folder at this level ("authorized folders")
+        #
+        ###
+        folder = op.split(root)[1]
+        folder_res = [re.compile(x).search(folder) is None
+                      for x in get_list_of_rules_with(currentdepth_rules['authorized_folders'])]
+
+        # if none of the authorized rules is respected, raise an error
+        if all(folder_res):
+            error_list.append("Naming rule not respected for this directory : {}".format(root))
+        ###
+        # 3. check whether rules are followed for files within the folder at this level ("authorized files")
+        ###
+        for current_file in files:
+            file_res = list()
+            for rules in currentdepth_rules['authorized_metadata_files']:
+                file_res.extend(([re.compile(x).search(current_file) is None for x in
+                                  get_list_of_rules_with(rules, Rs.AUTHORIZED_METADATA_EXTENSIONS)]))
+
+            for rules in currentdepth_rules['authorized_data_files']:
+                file_res.extend([re.compile(x).search(current_file) is None for x in
+                                 get_list_of_rules_with(rules, Rs.AUTHORIZED_DATA_EXTENSIONS)])
+                # if none of the authorized rules is respected, raise an error
+
+            if all(file_res):
+                error_list.append("Naming rule not respected for this file : {}".format(current_file))
+        ###
+        # 4. check whether the "mandatory files" are actually present at this level!
+        ###
+        if len(currentdepth_rules['mandatory_files']) > 0:
+            # loop over rules, each rule corresponding to one mandatory file
+            for current_mandatoryfolder_rule in (currentdepth_rules['mandatory_files']):
+                list_of_mandatory_folders = (get_list_of_rules_with
+                                             (current_mandatoryfolder_rule, current_mandatoryfolder_rule[1]))
+                for each_file in list_of_mandatory_folders:
+                    file_res = [re.compile(each_file).search(f) is None for f in files]
+                    if all(file_res):
+                        error_list.append("Mandatory file not found for this rule : {}".
+                                          format(current_mandatoryfolder_rule))
+
+    # if there are no errors, the data set is valid!
+    valid = len(error_list) == 0
+
+    return valid, error_list
+
+
+def get_list_of_rules_with(file_name_regexp, file_ext_regexp=None):
```

Comment:
I think the variable names used here could be more explicit. Since this function is rather generic, maybe also `build_rule_regexp` would be a better name? And `file_name_regexp` can actually contain more than just that, as it's a list of regexp_sets (is it actually restricted to filenames?).

Reply:
```file_name_regexp``` Can be both filenames or foldernames actually so yes it's a good idea to rename it

Reply:
Done